File: distrib/matlab/image_analysis/test_image_id.m
%TEST_IMAGE_ID  Compare low-rank image compression methods from librla.
%
%  Description
%  -----------
%    This script compares six low-rank approximation methods on RGB images:
%
%      1. Randomized SVD (svd_sketch)
%      2. Interpolative Decomposition (id_sketch)
%      3. Randomized SVD with oversampling and power iteration
%      4. ID with oversampling and power iteration
%      5. QR with column pivoting (qr_sketch)
%
%    The image is reshaped to a 2D matrix (m x 3n) for processing.
%
%    ID selects k skeleton columns and expresses remaining columns as
%    linear combinations: A(:, piv(k+1:end)) ≈ A(:, piv(1:k)) * T
%
%    Two-sided ID selects both skeleton columns and rows, storing only
%    a small core submatrix C = A(I,J) plus interpolation matrices.
%
%    Set use_single=true to run in single precision (faster, less memory).
%
%  Requirements
%  ------------
%    * librla.m in the MATLAB path
%    * Image file (see below)
%    * Octave users need the image package (loaded automatically)
%
%  See also: librla.id_sketch, librla.svd_sketch, librla.qr_sketch, imread
%
% ----------------------------------------------------------------------

if exist('OCTAVE_VERSION', 'builtin')
    pkg load image
end

%%A = imread('pexels-flickr-149387.jpg');
%%A = imread('pexels-anniroenkae-4793404.jpg');
A = imread('x1d-II-sample-02.jpg');

%%A = permute(A, [2 1 3]);

[m, n, nc] = size(A);
fprintf('Image size: %d x %d x %d\n', m, n, nc);

figure(1)
imagesc(A)
axis image
title('Original (RGB)')

k = 60;
use_single = false;  % set to true for single precision

if use_single
    conv = @single;
else
    conv = @double;
end

% Reshape RGB image to 2D matrix: m x (n*nc)
A2 = reshape(conv(A), m, n*nc);

%% ========================================================================
% Method 1: Randomized SVD for comparison
%% ========================================================================
fprintf('\n--- Randomized SVD ---\n');
tic;
[U, s, V] = librla.svd_sketch(A2, k);
B2_svd = U(:,1:k) * diag(s(1:k)) * V(:,1:k)';
toc

B_svd = reshape(B2_svd, m, n, nc);

figure(2)
imagesc(uint8(B_svd))
axis image
title(sprintf('Rank-%d randomized SVD', k))

rel_error_svd = norm(A2 - B2_svd, 'fro') / norm(A2, 'fro');
fprintf('SVD relative error: %.6e\n', rel_error_svd);

%% ========================================================================
% Method 2: Interpolative Decomposition (randomized)
%% ========================================================================
fprintf('\n--- Interpolative Decomposition (id_sketch) ---\n');
tic;
[k_id, piv, T] = librla.id_sketch(A2, k);
toc

fprintf('ID rank: %d\n', k_id);

% Reconstruct: skeleton columns + interpolated columns
% A(:, piv(1:k)) are skeleton columns (kept exactly)
% A(:, piv(k+1:end)) ≈ A(:, piv(1:k)) * T

% Build reconstruction in permuted order, then unpermute
skeleton = A2(:, piv(1:k_id));
interpolated = skeleton * T;

B2_id_perm = [skeleton, interpolated];  % columns in permuted order

% Unpermute to original column order
B2_id = zeros(size(A2), class(A2));
B2_id(:, piv(1:k_id)) = skeleton;
B2_id(:, piv(k_id+1:end)) = interpolated;

B_id = reshape(B2_id, m, n, nc);

figure(3)
imagesc(uint8(B_id))
axis image
title(sprintf('Rank-%d interpolative decomposition', k_id))

rel_error_id = norm(A2 - B2_id, 'fro') / norm(A2, 'fro');
fprintf('ID relative error: %.6e\n', rel_error_id);

%% ========================================================================
% Method 3: Randomized SVD with oversampling and power iteration
%% ========================================================================
extra = floor(0.5 * k);  % 50% oversampling
piter = 2;               % power iterations
fprintf('\n--- Randomized SVD (extra_samples=%d, power_iter=%d) ---\n', extra, piter);
tic;
[U3, s3, V3] = librla.svd_sketch(A2, k, 'extra_samples', extra, 'power_iter', piter);
B2_svd2 = U3(:,1:k) * diag(s3(1:k)) * V3(:,1:k)';
toc

B_svd2 = reshape(B2_svd2, m, n, nc);

figure(4)
imagesc(uint8(B_svd2))
axis image
title(sprintf('Rank-%d SVD (extra=%d, piter=%d)', k, extra, piter))

rel_error_svd2 = norm(A2 - B2_svd2, 'fro') / norm(A2, 'fro');
fprintf('SVD (oversampled+piter) relative error: %.6e\n', rel_error_svd2);

%% ========================================================================
% Method 4: Interpolative Decomposition with oversampling and power iteration
%% ========================================================================
fprintf('\n--- ID (extra_samples=%d, power_iter=%d) ---\n', extra, piter);
tic;
[k_id2, piv2, T2] = librla.id_sketch(A2, k, 'extra_samples', extra, 'power_iter', piter);
toc

fprintf('ID rank: %d\n', k_id2);

% Reconstruct
skeleton2 = A2(:, piv2(1:k_id2));
interpolated2 = skeleton2 * T2;

B2_id2 = zeros(size(A2), class(A2));
B2_id2(:, piv2(1:k_id2)) = skeleton2;
B2_id2(:, piv2(k_id2+1:end)) = interpolated2;

B_id2 = reshape(B2_id2, m, n, nc);

figure(5)
imagesc(uint8(B_id2))
axis image
title(sprintf('Rank-%d ID (extra=%d, piter=%d)', k_id2, extra, piter))

rel_error_id2 = norm(A2 - B2_id2, 'fro') / norm(A2, 'fro');
fprintf('ID (oversampled+piter) relative error: %.6e\n', rel_error_id2);

%% ========================================================================
% Method 5: QR with column pivoting (via qr_sketch)
%% ========================================================================
fprintf('\n--- QR (extra_samples=%d, power_iter=%d) ---\n', extra, piter);
tic;
[Q_qr, R_qr, p_qr] = librla.qr_sketch(A2, k, 'extra_samples', extra, 'power_iter', piter);
toc

k_qr = size(Q_qr, 2);
fprintf('QR rank: %d\n', k_qr);

% Reconstruct: A(:, p) ≈ Q * R, so unpermute columns
B2_qr_perm = Q_qr * R_qr;  % columns in permuted order
B2_qr = zeros(size(A2), class(A2));
B2_qr(:, p_qr) = B2_qr_perm;

B_qr = reshape(B2_qr, m, n, nc);

figure(6)
imagesc(uint8(B_qr))
axis image
title(sprintf('Rank-%d QR (extra=%d, piter=%d)', k_qr, extra, piter))

rel_error_qr = norm(A2 - B2_qr, 'fro') / norm(A2, 'fro');
fprintf('QR relative error: %.6e\n', rel_error_qr);

%% ========================================================================
% Summary
%% ========================================================================
fprintf('\n=== Summary ===\n');
fprintf('Method                              Rank    Relative Error\n');
fprintf('-----------------------------------------------------------\n');
fprintf('Randomized SVD                      %4d    %.6e\n', k, rel_error_svd);
fprintf('ID (randomized)                     %4d    %.6e\n', k_id, rel_error_id);
fprintf('SVD (extra=%d, piter=%d)            %4d    %.6e\n', extra, piter, k, rel_error_svd2);
fprintf('ID  (extra=%d, piter=%d)            %4d    %.6e\n', extra, piter, k_id2, rel_error_id2);
fprintf('QR  (extra=%d, piter=%d)            %4d    %.6e\n', extra, piter, k_qr, rel_error_qr);

File: distrib/image_analysis/test_image_id.py
"""
test_image_id.py - Compare low-rank image compression methods from librla.

Description
-----------
This script compares low-rank approximation methods on RGB images:

  1. Randomized SVD (svd_sketch)
  2. Interpolative Decomposition (id_sketch)
  3. Randomized SVD with oversampling and power iteration
  4. ID with oversampling and power iteration
  5. QR with column pivoting (qr_sketch)

The image is reshaped to a 2D matrix (m x 3n) for processing.

ID selects k skeleton columns and expresses remaining columns as
linear combinations: A[:, piv[k:]] = A[:, piv[:k]] @ T

Set use_single=True to run in single precision (faster, less memory).

Requirements
------------
* librla.py in the path
* Image file (see below)
* matplotlib, numpy, PIL

See also: librla.id_sketch, librla.svd_sketch, librla.qr_sketch

Author: Adrianna Gillman, Zydrunas Gimbutas
SPDX-License-Identifier: BSD-3-Clause
Version: 1.0.0
Date: TBD
Assisted by: Claude Code (Anthropic)
"""

import sys
import os
import time
import numpy as np
from numpy.linalg import norm
import matplotlib.pyplot as plt
from PIL import Image

# Add parent directory to path for librla
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))
import librla

# Load image
# image_file = 'pexels-flickr-149387.jpg'
image_file = 'pexels-anniroenkae-4793404.jpg'
A = np.array(Image.open(image_file))

# A = np.transpose(A, (1, 0, 2))

m, n, nc = A.shape
print(f'Image: {image_file}, size: {m} x {n} x {nc}')

plt.figure(1)
plt.imshow(A)
plt.title('Original (RGB)')
plt.draw()
plt.pause(0.1)

k = 60*2
use_single = True  # set to True for single precision

if use_single:
    conv = np.float32
else:
    conv = np.float64

# Reshape RGB image to 2D matrix: m x (n*nc)
A2 = conv(A).reshape(m, n * nc)

# ========================================================================
# Method 1: Randomized SVD for comparison
# ========================================================================
t0 = time.time()
U, s, Vh = librla.svd_sketch(A2, k)
B2_svd = U @ np.diag(s) @ Vh
elapsed_svd = time.time() - t0

B_svd = B2_svd.reshape(m, n, nc)

plt.figure(2)
plt.imshow(np.clip(B_svd, 0, 255).astype(np.uint8))
plt.title(f'Rank-{k} svd_sketch')
plt.draw()

rel_error_svd = norm(A2 - B2_svd, 'fro') / norm(A2, 'fro')
print(f'svd_sketch(k={k}): {elapsed_svd:.3f}s, error {rel_error_svd:.6e}')

# ========================================================================
# Method 2: Interpolative Decomposition (randomized)
# ========================================================================
t0 = time.time()
k_id, piv, T = librla.id_sketch(A2, k)
elapsed_id = time.time() - t0

# Reconstruct: skeleton columns + interpolated columns
# A[:, piv[:k]] are skeleton columns (kept exactly)
# A[:, piv[k:]] = A[:, piv[:k]] @ T

# Build reconstruction in permuted order, then unpermute
skeleton = A2[:, piv[:k_id]]
interpolated = skeleton @ T

# Unpermute to original column order
B2_id = np.zeros_like(A2)
B2_id[:, piv[:k_id]] = skeleton
B2_id[:, piv[k_id:]] = interpolated

B_id = B2_id.reshape(m, n, nc)

plt.figure(3)
plt.imshow(np.clip(B_id, 0, 255).astype(np.uint8))
plt.title(f'Rank-{k_id} id_sketch')
plt.draw()

rel_error_id = norm(A2 - B2_id, 'fro') / norm(A2, 'fro')
print(f'id_sketch(k={k_id}): {elapsed_id:.3f}s, error {rel_error_id:.6e}')

# ========================================================================
# Method 3: Randomized SVD with oversampling and power iteration
# ========================================================================
extra = k // 2  # 50% oversampling
piter = 2       # power iterations
t0 = time.time()
U3, s3, Vh3 = librla.svd_sketch(A2, k, extra_samples=extra, power_iter=piter)
B2_svd2 = U3 @ np.diag(s3) @ Vh3
elapsed_svd2 = time.time() - t0

B_svd2 = B2_svd2.reshape(m, n, nc)

plt.figure(4)
plt.imshow(np.clip(B_svd2, 0, 255).astype(np.uint8))
plt.title(f'Rank-{k} svd_sketch (extra_samples={extra}, power_iter={piter})')
plt.draw()

rel_error_svd2 = norm(A2 - B2_svd2, 'fro') / norm(A2, 'fro')
print(f'svd_sketch(k={k}, extra_samples={extra}, power_iter={piter}): {elapsed_svd2:.3f}s, error {rel_error_svd2:.6e}')

# ========================================================================
# Method 4: Interpolative Decomposition with oversampling and power iteration
# ========================================================================
t0 = time.time()
k_id2, piv2, T2 = librla.id_sketch(A2, k, extra_samples=extra, power_iter=piter)
elapsed_id2 = time.time() - t0

# Reconstruct
skeleton2 = A2[:, piv2[:k_id2]]
interpolated2 = skeleton2 @ T2

B2_id2 = np.zeros_like(A2)
B2_id2[:, piv2[:k_id2]] = skeleton2
B2_id2[:, piv2[k_id2:]] = interpolated2

B_id2 = B2_id2.reshape(m, n, nc)

plt.figure(5)
plt.imshow(np.clip(B_id2, 0, 255).astype(np.uint8))
plt.title(f'Rank-{k_id2} id_sketch (extra_samples={extra}, power_iter={piter})')
plt.draw()

rel_error_id2 = norm(A2 - B2_id2, 'fro') / norm(A2, 'fro')
print(f'id_sketch(k={k_id2}, extra_samples={extra}, power_iter={piter}): {elapsed_id2:.3f}s, error {rel_error_id2:.6e}')

# ========================================================================
# Method 5: QR with column pivoting (via qr_sketch)
# ========================================================================
t0 = time.time()
Q_qr, R_qr, p_qr = librla.qr_sketch(A2, k, extra_samples=extra, power_iter=piter)
elapsed_qr = time.time() - t0

k_qr = Q_qr.shape[1]

# Reconstruct: A[:, p] = Q @ R, so unpermute columns
B2_qr_perm = Q_qr @ R_qr  # columns in permuted order
B2_qr = np.zeros_like(A2)
B2_qr[:, p_qr] = B2_qr_perm

B_qr = B2_qr.reshape(m, n, nc)

plt.figure(6)
plt.imshow(np.clip(B_qr, 0, 255).astype(np.uint8))
plt.title(f'Rank-{k_qr} qr_sketch (extra_samples={extra}, power_iter={piter})')
plt.draw()

rel_error_qr = norm(A2 - B2_qr, 'fro') / norm(A2, 'fro')
print(f'qr_sketch(k={k_qr}, extra_samples={extra}, power_iter={piter}): {elapsed_qr:.3f}s, error {rel_error_qr:.6e}')

# ========================================================================
# Summary
# ========================================================================
print(f'\n{"Method":<40} {"Rank":>4}    {"Error"}')
print('-' * 55)
print(f'{"svd_sketch(k=" + str(k) + ")":<40} {k:4d}    {rel_error_svd:.6e}')
print(f'{"id_sketch(k=" + str(k_id) + ")":<40} {k_id:4d}    {rel_error_id:.6e}')
print(f'{"svd_sketch(k=" + str(k) + ", extra, power)":<40} {k:4d}    {rel_error_svd2:.6e}')
print(f'{"id_sketch(k=" + str(k_id2) + ", extra, power)":<40} {k_id2:4d}    {rel_error_id2:.6e}')
print(f'{"qr_sketch(k=" + str(k_qr) + ", extra, power)":<40} {k_qr:4d}    {rel_error_qr:.6e}')

plt.show(block=False)
